M2/Macaulay2/d/actors2.dd

```
@@ -425,96 +425,115 @@ export setStdIO   (f:file):void := ( stdIO    = f; setGlobalVariable(stdioS,  Ex
-     n := 0;
-     ff := openfiles;
-     while true do (
-	  when ff
-	  is null do break
-	  is f:FileCell do (n=n+1; ff=f.next;));
-     v := new Sequence len n do (
-	  ff = openfiles;
-	  while true do (
-	       when ff
-	       is null do break
-	       is f:FileCell do (provide f.file; ff=f.next;));
-	  );
-     list(v));
+    when e
+    is a:Sequence do (
+	-- typical value: openFiles, List
+	if length(a) == 0 then (
+	    n := 0;
+	    ff := openfiles;
+	    while true do (
+		when ff
+		is null do break
+		is f:FileCell do (n=n+1; ff=f.next;));
+	    v := new Sequence len n do (
+		ff = openfiles;
+		while true do (
+		    when ff
+		    is null do break
+		    is f:FileCell do (provide f.file; ff=f.next;));
+		);
+	    list(v))
+	else WrongNumArgs(0))
+    else WrongNumArgs(0));
 setupfun("openFiles",openfilesfun);
 openIn(filename:Expr):Expr := (
      when filename
+     -- # typical value: openIn, File, File
      is f:file do (
 	  when openIn(f)
 	  is g:file do Expr(g)
 	  is m:errmsg do buildErrorPacket(m.message))
+     -- # typical value: openIn, String, File
      is f:stringCell do (
 	  when openIn(f.v)
 	  is g:file do Expr(g)
 	  is m:errmsg do buildErrorPacket(m.message))
-     is Error do filename
-     else WrongArgString());
+     else WrongArg("a file or string"));
 setupfun("openIn",openIn);
 openOut(filename:Expr):Expr := (
      when filename
+     -- # typical value: openOut, File, File
      is f:file do (
 	  when openOut(f)
 	  is g:file do Expr(g)
 	  is m:errmsg do buildErrorPacket(m.message))
+     -- # typical value: openOut, String, File
      is f:stringCell do (
 	  when openOut(f.v)
 	  is g:file do Expr(g)
 	  is m:errmsg do buildErrorPacket(m.message))
-     is Error do filename
      else WrongArgString());
 setupfun("openOut",openOut);
 openOutAppend(filename:Expr):Expr := (
      when filename
+     -- # typical value: openOutAppend, File, File
+     is f:file do (
+	  when openOut(f)
```

Does `openOutAppend` make sense for listeners? Regardless, this is non-obvious enough it should probably have a comment before the call to `openOut(f)` that mentions that the only files this accepts are listeners

---

Hmm, good question.  Maybe we should just raise an error in this case?

---

I'd either have `openOutAppend` not accept listeners or do what you did here, but I'm leaning towards just adding a comment. But this detail should probably be documented. In particular the documentation openListener only mentions openInOut right now.

In reading through the networking stuff, my preferred solution for all of this is for accept to be a different method, and for `openListener` to return a Listener type which probably should be a subtype of File.

Right now `openIn`/`openOut`/`openInOut` on listeners all just call accept and so at the actual socket level it doesn't make a difference. The only difference is whether the Macaulay2 file object can be read/write from.